### PR TITLE
fix: Commit sled state in a transaction

### LINKED ISSUE #932: Panic in state service GetDepth request during obtain_tips
While I was syncing mainnet, I saw a panic in  GetDepth during obtain_tips:
<details>
<summary>Message:  block must be present if contains returned true<br/>
Location: zebra-state/src/on_disk.rs:166<br/>
<i>Click for full panic log.</i></summary>

```
Aug 20 14:51:37.588  INFO sync: zebrad::commands::start::sync: restarting sync e=Elapsed(())
Aug 20 14:51:37.588  INFO sync: zebrad::commands::start::sync: starting sync, obtaining new tips
Aug 20 14:51:37.589  INFO sync:obtain_tips: zebra_state: created block locator tip_height=Height(106000) min_locator_height=105901 locators=[Height(106000), Height(105999), Height(105998), Height(105996), Height(105992), Height(105984), Height(105968), Height(105936), Height(105901)]
Aug 20 14:51:48.626  INFO sync:obtain_tips:block_fetch_verify{hash=BlockHeaderHash("0e8fe552076141ce971a656a4ce3dd8e1b6100fe69b8e637c516951e00000000")}: zebra_consensus::checkpoint: Verified checkpoint range current_range=(Excluded(Height(106000)), Included(Height(107063)))
The application panicked (crashed).
Message:  block must be present if contains returned true
Location: zebra-state/src/on_disk.rs:166

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: zebrad::commands::start::sync::obtain_tips
      at zebrad/src/commands/start/sync.rs:162
   1: zebrad::commands::start::sync::sync
      at zebrad/src/commands/start/sync.rs:89
```

</details>

I think this panic happens when `AddBlock` and `GetDepth` are executed simultaneously, resulting in this order of events:
1. `AddBlock`: updates height tree with new height in `SledState::insert`
2. `GetDepth`: gets new hash using `SledState::get_tip`
3. `GetDepth`: gets None for new hash in `SledState::get`
4. `GetDepth`: expects that every tip hash will have a block in the hash tree, and panics

The `AddBlock` update to the hash tree in `SledState::insert` is not executed before step 3, but `GetDepth` expects the trees to be consistent, so it panics.

On relaunch, syncing restarts from the same checkpoint height. So it looks like the height tree update is never committed to the disk state, or sled rolls it back, due to the panic. (So the state isn't corrupted, which is good.)

### PR DESCRIPTION
**Pull request by teor2345**
Fixes a race condition between the height and hash tree `AddBlock` updates,
when they are executed simultaneously with GetDepth.

This is a temporary fix, until we implement serialised state reads.

Closes #932.

### CODE REVIEW

**Review by oxarbitrage — APPROVED** [review]
Code looks good to me. Did you tried to sync again with this patch and can confirm panics are gone ?

**Comment by teor2345** [comment]
> Did you tried to sync again with this patch and can confirm panics are gone ?

It seems to be a really rare race condition. I have never seen it before. Maybe some recent changes made it happen more often.

I haven't seen the panic again, but I am also having trouble getting sync to work reliably.

**Review by yaahc — APPROVED** [review]

File: zebra-state/src/on_disk.rs [hunk 1]
```diff
@@ -49,9 +51,11 @@ impl SledState {
...
         let bytes = block.zcash_serialize_to_vec()?;
 
-        // TODO(jlusby): make this transactional
-        height_map.insert(&height.0.to_be_bytes(), &hash.0)?;
-        by_hash.insert(&hash.0, bytes)?;
+        (&height_map, &by_hash).transaction(|(height_map, by_hash)| {
+            height_map.insert(&height.0.to_be_bytes(), &hash.0)?;
+            by_hash.insert(&hash.0, bytes.clone())?;
+            Ok(())
+        })?;
```
Comment (yaahc) [inline zebra-state/src/on_disk.rs:58]: huh, I swear this didn't compile last time I tried it, cool to see it just works now!

**Review by yaahc — APPROVED** [review]

File: zebra-state/src/on_disk.rs (hunk 1 above)
Comment (teor2345) [reply zebra-state/src/on_disk.rs:58]: The `use sled::Transactional;`, `bytes.clone()`, and the closure returning `Ok(())` were a little tricky to get right. But nightly rust gave me all sorts of good advice.

Comment (teor2345) [reply zebra-state/src/on_disk.rs:58]: Hmm also the from impl for the sled error.

**Pull request merged**